File: app/src/main/java/com/door43/translationstudio/ui/home/TargetTranslationListFragment.java
package com.door43.translationstudio.ui.home;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import com.door43.data.IPreferenceRepository;
import com.door43.translationstudio.R;
import com.door43.translationstudio.databinding.FragmentTargetTranslationListBinding;
import com.door43.translationstudio.ui.BaseFragment;
import com.door43.translationstudio.ui.viewmodels.HomeViewModel;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

/**
 * Displays a list of target translations
 */
@AndroidEntryPoint
public class TargetTranslationListFragment extends BaseFragment {

    @Inject
    IPreferenceRepository prefRepository;

    public static final String TAG = TargetTranslationListFragment.class.getSimpleName();
    public static final String STATE_SORT_BY_COLUMN = "state_sort_by_column";
    public static final String STATE_SORT_PROJECT_COLUMN = "state_sort_project_column";
    public static final String SORT_PROJECT_ITEM = "sort_project_item";
    public static final String SORT_BY_COLUMN_ITEM = "sort_by_column_item";

    private TargetTranslationAdapter adapter;
    private OnItemClickListener listener;
    private TargetTranslationAdapter.SortProjectColumnType sortProjectColumn = TargetTranslationAdapter.SortProjectColumnType.bibleOrder;
    private TargetTranslationAdapter.SortByColumnType sortByColumn = TargetTranslationAdapter.SortByColumnType.projectThenLanguage;

    private HomeViewModel viewModel;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        FragmentTargetTranslationListBinding binding =
                FragmentTargetTranslationListBinding.inflate(inflater, container, false);
        viewModel = new ViewModelProvider(requireActivity()).get(HomeViewModel.class);

        setupObservers();

        adapter = new TargetTranslationAdapter();
        adapter.setOnInfoClickListener(item -> {
            FragmentTransaction ft = getParentFragmentManager().beginTransaction();
            Fragment prev = getParentFragmentManager().findFragmentByTag("infoDialog");
            if (prev != null) {
                ft.remove(prev);
            }
            ft.addToBackStack(null);

            TargetTranslationInfoDialog dialog = new TargetTranslationInfoDialog();
            Bundle args = new Bundle();
            args.putString(TargetTranslationInfoDialog.ARG_TARGET_TRANSLATION_ID, item.getTranslation().getId());
            dialog.setArguments(args);
            dialog.show(ft, "infoDialog");
        });
        binding.translationsList.setAdapter(adapter);

        // open target translation
        binding.translationsList.setOnItemClickListener(
                (parent, view, position, id) -> listener.onItemClick(adapter.getItem(position))
        );

        if(savedInstanceState != null) {
            sortByColumn = TargetTranslationAdapter.SortByColumnType.fromInt(savedInstanceState.getInt(STATE_SORT_BY_COLUMN, sortByColumn.getValue()));
            sortProjectColumn = TargetTranslationAdapter.SortProjectColumnType.fromInt(savedInstanceState.getInt(STATE_SORT_PROJECT_COLUMN, sortProjectColumn.getValue()));
        } else { // if not restoring states, get last values
            sortByColumn = TargetTranslationAdapter.SortByColumnType.fromString(
                    prefRepository.getDefaultPref(SORT_BY_COLUMN_ITEM, null),
                    TargetTranslationAdapter.SortByColumnType.projectThenLanguage
            );
            sortProjectColumn = TargetTranslationAdapter.SortProjectColumnType.fromString(
                    prefRepository.getDefaultPref(SORT_PROJECT_ITEM, null),
                    TargetTranslationAdapter.SortProjectColumnType.bibleOrder
            );
        }
        adapter.sort(sortByColumn, sortProjectColumn);

        List<String> projectTypes = new ArrayList<>();
        projectTypes.add(this.getResources().getString(R.string.sort_project_then_language));
        projectTypes.add(this.getResources().getString(R.string.sort_language_then_project));
        projectTypes.add(this.getResources().getString(R.string.sort_progress_then_project));
        ArrayAdapter<String> projectTypesAdapter = new ArrayAdapter<>(requireActivity(), android.R.layout.simple_spinner_item, projectTypes);
        projectTypesAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        binding.sortColumn.setAdapter(projectTypesAdapter);
        binding.sortColumn.setSelection(sortByColumn.getValue());
        binding.sortColumn.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
//                    Logger.i(TAG, "Sort column item selected: " + position);
                sortByColumn = TargetTranslationAdapter.SortByColumnType.fromInt(position);
                prefRepository.getDefaultPref(SORT_BY_COLUMN_ITEM, String.valueOf(sortByColumn.getValue()));
                if(adapter != null) {
                    adapter.sort(sortByColumn, sortProjectColumn);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        List<String> bibleTypes = new ArrayList<>();
        bibleTypes.add(this.getResources().getString(R.string.sort_bible_order));
        bibleTypes.add(this.getResources().getString(R.string.sort_alphabetical_order));
        ArrayAdapter<String> bibleTypesAdapter = new ArrayAdapter<>(requireActivity(), android.R.layout.simple_spinner_item, bibleTypes);
        bibleTypesAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        binding.sortProjects.setAdapter(bibleTypesAdapter);
        binding.sortProjects.setSelection(sortProjectColumn.getValue());
        binding.sortProjects.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
//                    Logger.i(TAG, "Sort project column item selected: " + position);
                sortProjectColumn = TargetTranslationAdapter.SortProjectColumnType.fromInt(position);
                prefRepository.setDefaultPref(SORT_PROJECT_ITEM, String.valueOf(sortProjectColumn.getValue()));
                if(adapter != null) {
                    adapter.sort(sortByColumn, sortProjectColumn);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        return binding.getRoot();
    }

    private void setupObservers() {
        viewModel.getTranslations().observe(getViewLifecycleOwner(), translations -> {
            if (translations != null && adapter != null) {
                adapter.changeData(translations);
            }
        });
    }

    public void reloadList() {
        if (adapter != null) {
            adapter.notifyDataSetChanged();
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            this.listener = (OnItemClickListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context + " must implement OnItemClickListener");
        }
    }

    @Override
    public void onSaveInstanceState(Bundle out) {
        out.putInt(STATE_SORT_BY_COLUMN, sortByColumn.getValue());
        out.putInt(STATE_SORT_PROJECT_COLUMN, sortProjectColumn.getValue());
        super.onSaveInstanceState(out);
    }

    public interface OnItemClickListener {
        void onItemClick(TranslationItem item);
    }
}

File: app/src/main/java/com/door43/translationstudio/ui/home/TargetTranslationListFragment.kt
package com.door43.translationstudio.ui.home

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import androidx.fragment.app.activityViewModels
import com.door43.data.AssetsProvider
import com.door43.data.IPreferenceRepository
import com.door43.data.getDefaultPref
import com.door43.data.setDefaultPref
import com.door43.translationstudio.R
import com.door43.translationstudio.core.Typography
import com.door43.translationstudio.databinding.FragmentTargetTranslationListBinding
import com.door43.translationstudio.ui.BaseFragment
import com.door43.translationstudio.ui.dialogs.ProgressHelper
import org.koin.android.ext.android.inject

/**
 * Displays a list of target translations
 */
class TargetTranslationListFragment : BaseFragment() {
    val prefRepository: IPreferenceRepository by inject()
    val typography: Typography by inject()
    val assetsProvider: AssetsProvider by inject()

    //private var listener: OnItemClickListener? = null
    private var sortProjectColumn = BookSort.BibleOrder
    private var sortByColumn = ProjectSort.ProjectThenLanguage

    private var progressDialog: ProgressHelper.ProgressDialog? = null

    private lateinit var adapter: TargetTranslationAdapter
    private val viewModel: HomeViewModel by activityViewModels()

    private var _binding: FragmentTargetTranslationListBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentTargetTranslationListBinding.inflate(
            inflater,
            container,
            false
        )

        setupObservers()

        adapter = TargetTranslationAdapter(typography, assetsProvider)
        adapter.setOnInfoClickListener(object : TargetTranslationAdapter.OnInfoClickListener {
            override fun onClick(item: TranslationItem) {
                val ft = parentFragmentManager.beginTransaction()
                val prev = parentFragmentManager.findFragmentByTag("infoDialog")
                if (prev != null) {
                    ft.remove(prev)
                }
                ft.addToBackStack(null)

                val dialog = TargetTranslationInfoDialog()
                val args = Bundle()
                args.putString(
                    TargetTranslationInfoDialog.ARG_TARGET_TRANSLATION_ID,
                    item.translation.id
                )
                dialog.arguments = args
                dialog.show(ft, "infoDialog")
            }
        })
        binding.translationsList.adapter = adapter

        // open target translation
        binding.translationsList.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            //listener?.onItemClick(adapter.getItem(position))
        }

        if (savedInstanceState != null) {
            sortByColumn = ProjectSort.of(
                savedInstanceState.getInt(
                    STATE_SORT_BY_COLUMN,
                    sortByColumn.ordinal
                )
            )
            sortProjectColumn = BookSort.of(
                savedInstanceState.getInt(
                    STATE_SORT_PROJECT_COLUMN, sortProjectColumn.value
                )
            )
        } else { // if not restoring states, get last values
//            sortByColumn = ProjectSort.of(
//                prefRepository.getDefaultPref(SORT_BY_COLUMN_ITEM, 0)
//            )
//            sortProjectColumn = BookSort.of(
//                prefRepository.getDefaultPref(SORT_PROJECT_ITEM, 0)
//            )
        }
        adapter.sort(sortByColumn, sortProjectColumn)

        val projectTypes: MutableList<String> = ArrayList()
        projectTypes.add(this.resources.getString(R.string.sort_project_then_language))
        projectTypes.add(this.resources.getString(R.string.sort_language_then_project))
        projectTypes.add(this.resources.getString(R.string.sort_progress_then_project))
        val projectTypesAdapter =
            ArrayAdapter(requireActivity(), android.R.layout.simple_spinner_item, projectTypes)
        projectTypesAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.sortColumn.adapter = projectTypesAdapter
        binding.sortColumn.setSelection(sortByColumn.ordinal)
        binding.sortColumn.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
//                    Logger.i(TAG, "Sort column item selected: " + position);
                sortByColumn = ProjectSort.of(position)
                //prefRepository.getDefaultPref(SORT_BY_COLUMN_ITEM, sortByColumn.ordinal)
                adapter.sort(sortByColumn, sortProjectColumn)
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {
            }
        }

        val bibleTypes: MutableList<String> = ArrayList()
        bibleTypes.add(this.resources.getString(R.string.sort_bible_order))
        bibleTypes.add(this.resources.getString(R.string.sort_alphabetical_order))
        val bibleTypesAdapter =
            ArrayAdapter(requireActivity(), android.R.layout.simple_spinner_item, bibleTypes)
        bibleTypesAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.sortProjects.adapter = bibleTypesAdapter
        binding.sortProjects.setSelection(sortProjectColumn.value)
        binding.sortProjects.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
//                    Logger.i(TAG, "Sort project column item selected: " + position);
                sortProjectColumn = BookSort.of(position)
//                prefRepository.setDefaultPref(
//                    SORT_PROJECT_ITEM,
//                    sortProjectColumn.value.toString()
//                )
                adapter.sort(sortByColumn, sortProjectColumn)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {
            }
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        progressDialog = ProgressHelper.newInstance(
            childFragmentManager,
            R.string.loading,
            false
        )
    }

    private fun setupObservers() {

//        viewModel.translations.observe(viewLifecycleOwner) {
//            it?.let { adapter.setTranslations(it) }
//        }
    }

    fun reloadList() {
        val handler = Handler(Looper.getMainLooper())
        handler.post { adapter.notifyDataSetChanged() }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            //this.listener = context as OnItemClickListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement OnItemClickListener")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        progressDialog = null
    }

    override fun onSaveInstanceState(out: Bundle) {
        out.putInt(STATE_SORT_BY_COLUMN, sortByColumn.ordinal)
        out.putInt(STATE_SORT_PROJECT_COLUMN, sortProjectColumn.value)
        super.onSaveInstanceState(out)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        val TAG: String = TargetTranslationListFragment::class.java.simpleName
        const val STATE_SORT_BY_COLUMN: String = "state_sort_by_column"
        const val STATE_SORT_PROJECT_COLUMN: String = "state_sort_project_column"
    }
}
